chain/src/chain.rs

```diff
@@ -281,52 +244,223 @@ impl<CI: ChainIndex + 'static> ChainService<CI> {
             self.shared.store().insert_transaction_address(
                 batch,
                 &hash,
-                &block.commit_transactions(),
+                block.commit_transactions(),
             );
         }
 
-        let mut hash = block.header().parent_hash().clone();
-        number -= 1;
+        for n in new_number..old_number {
+            self.shared.store().delete_block_hash(batch, n + 1);
+        }
+    }
+
+    fn get_forks(
+        &self,
+        tip_number: BlockNumber,
+        block: &Block,
+        exts: &mut Vec<(H256, BlockExt)>,
+        ext: BlockExt,
+    ) -> (bool, Vec<Block>, Vec<Block>) {
+        let mut number = block.header().number() - 1;
+        let mut old_blocks = Vec::new();
+        let mut new_blocks = Vec::new();
+
+        // The old fork may longer than new fork
+        if number < tip_number {
+            for n in number..=tip_number {
+                let hash = self.shared.block_hash(n).unwrap();
+                let old_block = self.shared.block(&hash).unwrap();
+
+                old_blocks.push(old_block);
+            }
+        }
+
+        let mut verified = ext.valid.is_some();
+        let mut hash = block.header().hash();
+
+        if !verified {
+            exts.push((hash, ext));
+        }
+
+        //TODO: remove this clone
+        new_blocks.push(block.clone());
+
+        hash = block.header().parent_hash().clone();
+
         loop {
             if let Some(old_hash) = self.shared.block_hash(number) {
```

Comment:
Better to add an assert/expect here. If number is less than or equal to tip_number, it must return a hash. If number is larger than tip_number, we don't need to bother looking up the hash from the store.

Reply:
Block number is not sufficient to estimate which chain is better. It is easy to construct a chain which is longer but has much lower difficulty.

Reply:
My point is:

- If `number > tip_number`, it must be absent, we don't need to look up it in the store.
- If `number <= tip_number`, the store must return its hash, we should assert it.